addons/mail/models/ir_binary.py

```diff
@@ -0,0 +1,18 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+
+class IrBinary(models.AbstractModel):
+    _inherit = ["ir.binary"]
+
+    def _find_record_check_access(self, record, access_token, field):
+        if record._name in ["res.partner", "mail.guest"] and field == "avatar_128":
+            # access to the avatar is allowed if there is access to the messages
+            if record._name == "res.partner":
+                domain = [("author_id", "=", record.id)]
+            else:
+                domain = [("author_guest_id", "=", record.id)]
+            if self.env["mail.message"].search_count(domain, limit=1):
+                return record.sudo()
+        return super()._find_record_check_access(record, access_token, field)
```

Comment:
So if I understand the fix correctly, this code means that I could see avatar of anyone as long as I can see any of this partner/guest messages?

It works most of the time... But if you don't want to show your partner avatar in some contexts then this would be accessible. In livechat we have "Online Chat Name" to have an operator name that's different from partner name. There's no way at the moment to set an "Online Chat Avatar"... but this feature would make sense too.

The feature can still exist with this code but at the cost of some privacy concerns. There's no need with current set of features in Discuss and Livechat, but this is probably a concern in near future.

Reply:
Yes, and more. If you're a member of a public channel, the avatar is also accessible. We're aware of this, and a solution like you mentioned could be to have two types of avatars (private and public).
Considering that most people use the same avatar almost everywhere, and that there's no force to use the actual image for internal users, we thought it would be fine for now.

Reply:
Did you really add a search on mail.message (most complex searchable model in odoo) for each avatar to display ? Wasn't it possible to find another solution based directly on displayed mail.message itself ?

Side question: why 580 tests lines for a 18 lines fix ?

Reply:
Preliminary note: the method will be improved in master (draft PR https://github.com/odoo/odoo/pull/186003) to only be called when strictly necessary, so only if ACL check fails.

So pretty much it will never be called for internal users, never in common channels (as the channel condition comes first, that is already the case in 18.0 too), etc.

If we have a performance issue, we might back-port the "checking ACL first" part in 18.0 too, that would make a lot of sense and avoid any query most of the time.
Do you have a reported issue or a benchmark related to this change? We can always have a look if necessary.

As for reasoning, the code is much cleaner this way than managing access through specific message, having to de-duplicating routes, losing ability to cache images due to using different routes, juggling with many routes depending on context in client code, etc.

As for the number of tests, it's just because there are that many cases to cover. We have at least 6 different types of "user profile" to cover, each on many different documents that each implement their own way of accessing the document/its messages.

Reply:
Well this is clearly overengineered, and for perf issues, there is already https://github.com/odoo/odoo/pull/187072 . It shows this adds queries and ACL check to access ... an avatar. Sometimes you have to make things simpler, and consider simple rules like "an avatar is always accessible".

For tests, well I removed about 25% of them that were complete no-brain duplicates. Anyway there is 0 comments, 0 explanations, and most of them are just copy-pasted: why testing with internal user, demo user then admin ? Only in some cases where admin has privilege rights should be tested, otherwise it is just triple testing the same feature.

Consider simple code and solutions is part of the development, doing 1K LOC to display avatars is something one might consider a bit complicated.

Cheers,

Reply:
We never know when devs put conditions on `_is_internal` or `_is_admin` or even checking groups manually in the code (one way or another: to add or to restrict access) effectively messing up with ACL. Then suddenly admin and internal users don't work the same anymore.

I can write very simple "code mistake" that makes the "full test suite" as it was written properly fail, but for which the "no-duplicate" naive suite incorrectly passes. So I personally prefer to test everything rather than assuming it works because it should, especially on security matters.

I'm not against the cleanup that was done to the tests, I think it stayed in the reasonable range, but I'd be afraid to reduce the coverage further than that though, and I'd try to keep at least one relevant test with the full coverage.

Reply:
As for the performance issue, avatar highlighted it that's true, but it's an issue we had generally speaking with the introduction of sub channel. And the channel part wasn't even added in this PR, this was only about messages here.

And we still need to merge/back-port the checking of ACL before checking specific stuff, then it won't ever do the extra code for internal users and we only get the queries in rare cases, when we actually need them (as opposed to if we do a channel route, we always have the queries in channels for example + losing caching + a lot of other drawbacks).

Reply:
Yeah absolutely no-one told you it could have been a good idea to simplify :roll_eyes:  . Not saying all refactoring are useless, but please consider that KISS is a real guidelines in python (Keep It Super Simple).